example/echo-op/echo_op.cpp

```diff
@@ -97,132 +97,114 @@ async_echo (
-    // CompletionToken, captures a local reference to the handler,
-    // and creates the `async_result` object which becomes the
-    // return value of this initiating function.
+    // If a newline is present in the buffer sequence, this function returns
+    // the number of characters from the beginning of the buffer up to the
+    // newline, including the newline character. Otherwise it returns zero.
 
-    net::async_completion<CompletionToken, void(beast::error_code)> init(token);
+    std::size_t
+    find_newline(typename DynamicBuffer::const_buffers_type const& buffers)
+    {
+        // The `buffers_iterator` class template provides random-access
+        // iterators into a buffer sequence. Use the standard algorithm
+        // to look for the new line if it exists.
 
-    // The helper macro BOOST_ASIO_HANDLER_TYPE converts the completion
-    // token type into a concrete handler type of the correct signature.
+        auto begin = net::buffers_iterator<
+            typename DynamicBuffer::const_buffers_type>::begin(buffers);
+        auto end =   net::buffers_iterator<
+            typename DynamicBuffer::const_buffers_type>::end(buffers);
+        auto result = std::find(begin, end, '\n');
 
-    using handler_type = BOOST_ASIO_HANDLER_TYPE(CompletionToken, void(beast::error_code));
+        if(result == end)
+            return 0; // not found
 
-    // The class template `async_base` holds the caller's completion
-    // handler for us, and provides all of the boilerplate for forwarding
-    // the associated allocator and associated executor from the caller's
-    // handler to our operation. It also maintains a `net::executor_work_guard`
-    // for the executor associated with the stream. This work guard is
-    // inexpensive, and prevents the execution context from running out
-    // of work. It is usually necessary although rarely it can be skipped
-    // depending on the operation (this echo example needs it because it
-    // performs more than one asynchronous operation in a row).
-    // We declare this type alias to make the code easier to read.
+        return result + 1 - begin;
+    }
+
+    // A type to differentiate the source of an invocation of Self
 
-    using base_type = beast::async_base<
-        handler_type, /*< The type of the completion handler obtained from the token >*/
-        beast::executor_type<AsyncStream> /*< The type of executor used by the stream to dispatch asynchronous operations >*/
-    >;
+    struct completion_tag {};
 
-    // This nested class implements the echo composed operation as a
-    // stateful completion handler. We derive from `async_base` to
-    // take care of boilerplate and we derived from asio::coroutine to
-    // allow the reenter and yield keywords to work.
+    // It is a requirement of async operations that the completion handler must be
+    // invoked as if by @ref post. The composed operation must know whether it is
```

Comment:
`post` is not a Beast function so making it a `@ref` won't work. Use `net::post` instead of `@ref post`.